source/beamline/CLS/CLSJJSlitsBladesControl.cpp

```diff
@@ -0,0 +1,339 @@
+			connect( lowerBladeControl_, SIGNAL(alarmChanged(int,int)), this, SIGNAL(alarmChanged(int,int)) );
+			connect( lowerBladeControl_, SIGNAL(error(int)), this, SIGNAL(error(int)) );
+		}
+
+		updateControlStates();
+
+		emit lowerBladeControlChanged(lowerBladeControl_);
+	}
+}
+
+void CLSJJSlitsBladesControl::setConnected(bool isConnected)
+{
+	if (connected_ != isConnected) {
+		connected_ = isConnected;
+		emit connected(connected_);
+	}
+}
+
+void CLSJJSlitsBladesControl::setValue(double newValue)
+{
+	if (value_ != newValue) {
+		value_ = newValue;
+		emit valueChanged(value_);
+	}
+}
+
+void CLSJJSlitsBladesControl::setSetpoint(double newValue)
+{
+	if (setpoint_ != newValue) {
+		setpoint_ = newValue;
+		emit setpointChanged(setpoint_);
+	}
+}
+
+void CLSJJSlitsBladesControl::setMoveInProgress(bool isMoving)
+{
+	if (moveInProgress_ != isMoving) {
+		moveInProgress_ = isMoving;
+		emit movingChanged(moveInProgress_);
+	}
+}
+
+void CLSJJSlitsBladesControl::setGap(double newValue)
+{
+	if (gap_ != newValue) {
+		gap_ = newValue;
+		emit gapChanged(gap_);
+	}
+}
+
+void CLSJJSlitsBladesControl::setCenterPosition(double newValue)
+{
+	if (centerPosition_ != newValue) {
+		centerPosition_ = newValue;
+		emit centerPositionChanged(centerPosition_);
+	}
+}
+
+void CLSJJSlitsBladesControl::updateControlStates()
+{
+	updateConnected();
+	updateGap();
+	updateCenterPosition();
+	updateValue();
+}
+
+void CLSJJSlitsBladesControl::updateConnected()
+{
+	bool upperControlOK = (upperBladeControl_ && upperBladeControl_->isConnected());
+	bool lowerControlOK = (lowerBladeControl_ && lowerBladeControl_->isConnected());
+
+	bool connected = (upperControlOK && lowerControlOK);
+
+	if (connected)
+		qDebug() << name() << "connected.";
+	else
+		qDebug() << name() << "NOT connected.";
+
+	setConnected(connected);
+}
+
+//void CLSJJSlitsBladesControl::updateValue()
+//{
+//	qDebug() << "BladesControl::updateValue()";
+//}
+
+void CLSJJSlitsBladesControl::updateGap()
+{
+	if (isConnected()) {
+		setGap( calculateGap(upperBladeControl_->value(), lowerBladeControl_->value()) );
+	}
+}
+
+void CLSJJSlitsBladesControl::updateCenterPosition()
+{
+	if (isConnected()) {
+		setCenterPosition( calculateCenterPosition(upperBladeControl_->value(), lowerBladeControl_->value()) );
+	}
+}
+
+void CLSJJSlitsBladesControl::onMoveStarted()
+{
+	setMoveInProgress(true);
+	emit moveStarted();
+}
+
+void CLSJJSlitsBladesControl::onMoveCancelled(QObject *action)
+{
+	moveCleanup(action);
+	emit moveFailed(AMControl::WasStoppedFailure);
+}
+
+void CLSJJSlitsBladesControl::onMoveFailed(QObject *action)
+{
+	moveCleanup(action);
+	emit moveFailed(AMControl::OtherFailure);
+}
+
+void CLSJJSlitsBladesControl::onMoveSucceeded(QObject *action)
+{
+	moveCleanup(action);
+	emit moveSucceeded();
+}
+
+void CLSJJSlitsBladesControl::moveCleanup(QObject *action)
+{
+	if (action) {
+		setMoveInProgress(false);
+
+		disconnect( action, 0, moveCancelled_, 0 );
```

Comment:
You could replace these three lines with ```action->disconnect();```.  With no arguments, that will remove all connections.  If you don't want to remove all the connections then you can leave this as is since it is more explicit.